internal/build/phase_config_provider.go

```diff
@@ -0,0 +1,136 @@
+		name:     name,
+	}
+
+	provider.ctrConf.Cmd = []string{"/cnb/lifecycle/" + name}
+	provider.ctrConf.Image = lifecycle.builder.Name()
+	provider.ctrConf.Labels = map[string]string{"author": "pack"}
+
+	ops = append(ops,
+		WithLifecycleProxy(lifecycle),
+		WithBinds([]string{
+			fmt.Sprintf("%s:%s", lifecycle.LayersVolume, layersDir),
+			fmt.Sprintf("%s:%s", lifecycle.AppVolume, appDir),
+		}...),
+	)
+
+	for _, op := range ops {
+		if err := op(provider); err != nil {
+			return nil, errors.Wrap(err, "create phase config")
+		}
+	}
+
+	return provider, nil
+}
+
+func (p *PhaseConfigProvider) ContainerConfig() *container.Config {
+	return p.ctrConf
+}
+
+func (p *PhaseConfigProvider) HostConfig() *container.HostConfig {
+	return p.hostConf
+}
+
+func (p *PhaseConfigProvider) Name() string {
+	return p.name
+}
+
+func (p *PhaseConfigProvider) Update(ops ...PhaseConfigProviderOperation) error {
+	for _, op := range ops {
+		if err := op(p); err != nil {
+			return errors.Wrap(err, "update phase config")
+		}
+	}
+	return nil
+}
+
+func WithArgs(args ...string) PhaseConfigProviderOperation {
+	return func(provider *PhaseConfigProvider) error {
+		provider.ctrConf.Cmd = append(provider.ctrConf.Cmd, args...)
+		return nil
+	}
+}
+
+func WithBinds(binds ...string) PhaseConfigProviderOperation {
+	return func(provider *PhaseConfigProvider) error {
+		provider.hostConf.Binds = append(provider.hostConf.Binds, binds...)
+		return nil
+	}
+}
+
+func WithDaemonAccess() PhaseConfigProviderOperation {
+	return func(provider *PhaseConfigProvider) error {
+		provider.ctrConf.User = "root"
+		provider.hostConf.Binds = append(provider.hostConf.Binds, "/var/run/docker.sock:/var/run/docker.sock")
+		return nil
+	}
+}
+
+func WithLifecycleProxy(lifecycle *Lifecycle) PhaseConfigProviderOperation {
+	return func(provider *PhaseConfigProvider) error {
+		if lifecycle.httpProxy != "" {
+			provider.ctrConf.Env = append(provider.ctrConf.Env, "HTTP_PROXY="+lifecycle.httpProxy)
+			provider.ctrConf.Env = append(provider.ctrConf.Env, "http_proxy="+lifecycle.httpProxy)
+		}
+
+		if lifecycle.httpsProxy != "" {
+			provider.ctrConf.Env = append(provider.ctrConf.Env, "HTTPS_PROXY="+lifecycle.httpsProxy)
+			provider.ctrConf.Env = append(provider.ctrConf.Env, "https_proxy="+lifecycle.httpsProxy)
+		}
+
+		if lifecycle.noProxy != "" {
+			provider.ctrConf.Env = append(provider.ctrConf.Env, "NO_PROXY="+lifecycle.noProxy)
+			provider.ctrConf.Env = append(provider.ctrConf.Env, "no_proxy="+lifecycle.noProxy)
+		}
+
+		return nil
+	}
+}
+
+func WithNetwork(networkMode string) PhaseConfigProviderOperation {
+	return func(provider *PhaseConfigProvider) error {
+		provider.hostConf.NetworkMode = container.NetworkMode(networkMode)
+		return nil
+	}
+}
+
+func WithRegistryAccess(repos ...string) PhaseConfigProviderOperation {
+	return func(provider *PhaseConfigProvider) error {
+		authConfig, err := auth.BuildEnvVar(authn.DefaultKeychain, repos...)
+		if err != nil {
+			return err
```

Comment:
Our operation method signature is required to return an error for every type of operation, where this is the only place in our seven current methods that we could actually return an error. If we could find a way to move this logic out of this operation we could simplify every other operation implementation.

Reply:
This seems reasonable and remove a bunch of error checking from `phases.go`.